### PR TITLE
import from gir

### PR DESCRIPTION
**Pull request by vhdirk**
This is what's generated with gir. All tests pass.

### CODE REVIEW

File: gio-subclass/src/application.rs [hunk 1]
```diff
@@ -76,104 +96,86 @@ impl convert::Into<Vec<OsString>> for ArgumentList {
...
 
-    fn open(&self, application: &T, files: &[gio::File], hint: &str) {
-        application.parent_open(files, hint)
+    fn before_emit(&self, application: &T, platform_data: &glib::Variant){
+        application.parent_before_emit(platform_data)
     }
 
-    fn command_line(&self, application: &T, cmd_line: &gio::ApplicationCommandLine) -> i32 {
-        application.parent_command_line(cmd_line)
+    fn command_line(&self, application: &T, command_line: /*Ignored*/&gio::ApplicationCommandLine) -> i32{
```
Comment (sdroege) [inline gio-subclass/src/application.rs:112]: What is the ignored comment here and in various other places?

**Review by sdroege — COMMENTED** [review]
This looks great :) Which part is exactly autogenerated? All of it without any changes whatsoever, including the `ArgumentList` type?

I would've expected that one to be custom code :)


Generally for the code organization, we usually put custom things into src, autogenerated things into src/auto (and re-export from src/lib.rs). And maybe the custom vfunc/etc snippets here could be put into custom at the top-level

File: gio-subclass/src/application.rs (hunk 1 above)
Comment (vhdirk) [reply gio-subclass/src/application.rs:112]: Not sure. I've reused as much as I could from other places. These comments where inserted as a result of that. I didn't really look into the 'why' of it all.

**Comment by vhdirk** [comment]
This actually includes a fair bit of custom inserted code. Pretty much everything related to ArgumentList, and then some...
I've included those bits in the gir PR, for now. See https://github.com/vhdirk/gir/tree/subclass/tests/subclass/gio-subclass. There it is also more inline with the de-facto default repo structure.

File: gio-subclass/src/application.rs (hunk 1 above)
Comment (sdroege) [reply gio-subclass/src/application.rs:112]: Ok, let's keep a note and look into that at a later time then :)

**Comment by sdroege** [comment]
> This actually includes a fair bit of custom inserted code. Pretty much everything related to ArgumentList, and then some...
> I've included those bits in the gir PR, for now. See https://github.com/vhdirk/gir/tree/subclass/tests/subclass/gio-subclass. There it is also more inline with the de-facto default repo structure.

It would seem better to have this all included in this repository, so that the autogenerated code can be regenerated easily whenever something is changing in the code generator

**Comment by sdroege** [comment]
This looks awesome :)

**Comment by vhdirk** [comment]
From now on, generating these bindings should be relatively easy :)

**Comment by vhdirk** [comment]
Is this good enough to merge? Then I can rebase the gaction branch

**Comment by sdroege** [comment]
Probably, I'll try to go over it during the weekend but generally it seems great :)

File: gio-subclass/src/application.rs (hunk 1 above)
Comment (sdroege) [reply gio-subclass/src/application.rs:112]: Can you create an issue for that after merging this?

**Comment by sdroege** [comment]
Looks good to go to me. Is there some kind of `Gir.toml` file for configuring which vfuncs to ignore, which to take from elsewhere, for overriding other things? Like for all the other modes of gir. If so, please add that to the repository too. Otherwise there should probably be one: for white-listing what to generate and for all kinds of overrides :)

Anyway, apart from that please merge

**Comment by vhdirk** [comment]
Yes. There is a gir.toml. It currently resides in the gir PR. I'll add it here too. Any preferred location?

**Comment by sdroege** [comment]
Root of the repository, `Gir_Gio.toml`

**Comment by vhdirk** [comment]
There it is. It does contain a (now) invalid path to the gir files...

**Comment by sdroege** [comment]
Looks fine. Maybe just add the required gir-files into a subdirectory here, like in gstreamer-rs. But good to go then, please merge :)

**Comment by vhdirk** [comment]
What about using submodules like in https://github.com/gtk-rs/gio?files=1

**Comment by sdroege** [comment]
I dislike git submodules, but maybe we should mirror what gtk-rs does here. So go ahead, add a submodule for this :)

**Comment by vhdirk** [comment]
Me too, but I dislike making copies even more :). And I have yet to find a decent alternative to submodules (and don't get me started on git subtrees :) ). Until there is one, submodules'll have to do.

**Comment by sdroege** [comment]
Thanks a lot, especially for your patience to get the autogeneration that far :)

**Comment by vhdirk** [comment]
No problem. Though I am now officially fed up with the generator :)
I actually need gtkapplicationwindow. Seeing how much layers I had to go through was the motivation to work on gir.
Final note though: how would you override interface vfuncs when subclassing, for instance, application?

**Comment by sdroege** [comment]
You would implement the interface in your subclass. If the parent class also implements the interface then this overrides the parent implementation IIRC.

That's what you mean?

**Comment by vhdirk** [comment]
Yes. Going to try it out tomorrow.

**Comment by sdroege** [comment]
This was breaking `gtk-subclass`: https://travis-ci.org/sdroege/gtk-subclass/jobs/396052127#L789-L798

Using the correct macro name (`box_gio_application_impl`) gives
```
error[E0277]: the trait bound `std::boxed::Box<application::GtkApplicationImpl<application::Application> + 'static>: application::GtkApplicationImpl<application::Application>` is not satisfied
  --> gtk-subclass/src/application.rs:95:13
   |
95 | unsafe impl GtkApplicationClassExt<Application> for ApplicationClass {}
   |             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ the trait `application::GtkApplicationImpl<application::Application>` is not implemented for `std::boxed::Box<application::GtkApplicationImpl<application::Application> + 'static>`

```

Didn't look closer yet :)

**Comment by vhdirk** [comment]
See https://github.com/sdroege/gtk-subclass/pull/6. I have it fixed there.

**Pull request merged**